source/light/contracts/Light.sol

```diff
@@ -93,51 +104,65 @@ contract Light is ILight {
    * @param signerAmount Amount for signerToken
    * @param senderToken Contract address of the ERC20 token that will be transferred from the sender
    * @param senderAmount Amount for senderToken
-   * @param signature Signature generated by the signer
    */
   function swap(
     uint256 nonce,
     uint256 expiry,
+    address signerWallet,
     IERC20 signerToken,
     uint256 signerAmount,
     IERC20 senderToken,
     uint256 senderAmount,
-    bytes calldata signature
+    uint8 v,
+    bytes32 r,
+    bytes32 s
   ) external override {
     require(DOMAIN_CHAIN_ID == getChainId(), "CHAIN_ID_CHANGED");
 
     // Ensure the expiry has not passed.
     require(expiry > block.timestamp, "EXPIRY_PASSED");
 
     bytes32 hashed =
-      _getHash(
+      _getOrderHash(
         nonce,
         expiry,
+        signerWallet,
         signerToken,
         signerAmount,
+        msg.sender,
         senderToken,
         senderAmount
       );
 
     // Recover the signer from the hash and signature.
-    address signer = _getSigner(hashed, signature);
+    address signer = _getSigner(hashed, v, r, s);
 
     // Ensure the nonce is above the minimum.
     require(nonce >= signerMinimumNonce[signer], "NONCE_TOO_LOW");
 
     // Mark the nonce as used and ensure it hasn't been used before.
     require(_markNonceAsUsed(signer, nonce), "NONCE_ALREADY_USED");
 
+    if (signerWallet != signer) {
+      require(authorized[signerWallet] == signer, "UNAUTHORIZED");
+    }
+
     // Transfer token from sender to signer.
-    senderToken.safeTransferFrom(msg.sender, signer, senderAmount);
+    senderToken.safeTransferFrom(msg.sender, signerWallet, senderAmount);
 
     // Transfer token from signer to sender.
-    signerToken.safeTransferFrom(signer, msg.sender, signerAmount);
+    signerToken.safeTransferFrom(signerWallet, msg.sender, signerAmount);
```

Comment:
You should add another method that allows sender to configure the destination. This can reduce the total number of transfers required to get funds to the real sender.

Reply:
good idea